File: human/node/game.js
import { spawn } from "child_process";
import { PATH, PORT, VERSION } from "./config.js";

export default class Game {

  async start(port) {
    console.log("Starting StarCraft II...");

    this.process = spawn("..\\Versions\\" + VERSION + "\\SC2_x64.exe", [
      "-dataVersion", "B89B5D6FA7CBF6452E721311BFBC6CB2",
      "-displaymode", "0", "-windowx", "0", "-windowy", "0", "-windowwidth", "800", "-windowheight", "500",
      "-listen", "127.0.0.1", "-port", String(port || PORT)
    ], {
      cwd: PATH + "\\Support64",
      stdio: "inherit",
    });

    this.process.on("error", this.exit.bind(this));
    this.process.on("exit", this.exit.bind(this));
  }

  exit(details) {
    console.error("StarCraft II stopped", details || "");
    this.process = null;
  }

  stop() {
    if (this.process) this.process.kill();
  }

}

File: human/rust/src/game.rs
use anyhow::Result;
use std::process::Stdio;
use std::sync::Arc;
use tokio::process::{Child, Command};

use crate::{config::Config, lobby::get_sc2_version_path};

pub struct Game {
    config: Arc<Config>,
    process: Option<Child>,
}

impl Game {
    pub fn new(config: Arc<Config>) -> Self {
        Self {
            config,
            process: None,
        }
    }

    pub async fn start(&mut self) -> Result<()> {
        println!("Starting StarCraft II...");

        let cwd = self.config.sc2_path.join("Support64");
        let exe = get_sc2_version_path(&self.config);
        // Note this doesn't change from linux to windows
        // But later config may prove that wine is required
        let child =  if cfg!(target_os = "windows"){
            Command::new(&exe)
                .args([
                    "-dataVersion",
                    "B89B5D6FA7CBF6452E721311BFBC6CB2",
                    "-displaymode",
                    "1",
                    "-listen",
                    &self.config.local_host,
                    "-port",
                    &self.config.sc2_port.to_string(),
                ])
                .current_dir(&cwd)
                .stdin(Stdio::inherit())
                .stdout(Stdio::inherit())
                .stderr(Stdio::inherit())
                .spawn()?
        } else if cfg!(target_os = "linux"){
             Command::new(&exe)
                .args([
                    "-dataVersion",
                    "B89B5D6FA7CBF6452E721311BFBC6CB2",
                    "-displaymode",
                    "1",
                    "-listen",
                    &self.config.local_host,
                    "-port",
                    &self.config.sc2_port.to_string(),
                ])
                .current_dir(&cwd)
                .stdin(Stdio::inherit())
                .stdout(Stdio::inherit())
                .stderr(Stdio::inherit())
                .spawn()?
        } else {
            unreachable!("Unsupported OS")
        };
        self.process = Some(child);
            Ok(())
    }

    pub async fn stop(&mut self) {
        if let Some(child) = &mut self.process {
            let _ = child.kill().await;
            let _ = child.wait().await;
        }
        self.process = None;
    }
}
